lib/seahorn/BvOpsem2WideMemMgr.cc

```diff
@@ -0,0 +1,543 @@
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    RawPtrTy getRaw() { return strct::extractVal(m_v, 0); }
+
+    RawPtrTy getSize() { return strct::extractVal(m_v, 1); }
+  };
+
+  struct MemValTyImpl {
+    Expr m_v;
+
+    MemValTyImpl(RawMemValTy &&val, Expr &&size) {
+      m_v = strct::mk(std::move(val), std::move(size));
+    }
+
+    MemValTyImpl(const RawPtrTy &val, const Expr &size) {
+      m_v = strct::mk(val, size);
+    }
+
+    explicit MemValTyImpl(const Expr &e) {
+      // Our ptr is a struct of two exprs
+      assert(strct::isStructVal(e));
+      assert(e->arity() == 2);
+      m_v = e;
+    }
+
+    Expr v() const { return m_v; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    RawMemValTy getRaw() { return strct::extractVal(m_v, 0); }
+
+    RawMemValTy getSize() { return strct::extractVal(m_v, 1); }
+  };
+
+  struct PtrSortTyImpl {
+    Expr m_ptr_sort;
+
+    PtrSortTyImpl(RawPtrSortTy &&ptr_sort, Expr &&size_sort) {
+      m_ptr_sort = sort::structTy(std::move(ptr_sort), std::move(size_sort));
+    }
+
+    PtrSortTyImpl(const RawPtrSortTy &ptr_sort, const Expr &size) {
+      m_ptr_sort = sort::structTy(ptr_sort, size);
+    }
+
+    Expr v() const { return m_ptr_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    RawPtrSortTy getRawSort() { return m_ptr_sort->arg(0); }
+  };
+
+  struct MemSortTyImpl {
+    Expr m_mem_sort;
+
+    MemSortTyImpl(RawMemSortTy &&mem_sort, Expr &&size_sort) {
+      m_mem_sort = sort::structTy(std::move(mem_sort), std::move(size_sort));
+    }
+
+    MemSortTyImpl(const RawMemSortTy &mem_sort, const Expr &size_sort) {
+      m_mem_sort = sort::structTy(mem_sort, size_sort);
+    }
+
+    Expr v() const { return m_mem_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
+  using PtrTy = PtrTyImpl;
+  using PtrSortTy = PtrSortTyImpl;
+  using MemValTy = MemValTyImpl;
+  using MemSortTy = MemSortTyImpl;
+
+  /// \brief A null pointer expression (cache)
+  PtrTy m_nullPtr;
+
+  WideMemManager(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned ptrSz,
+                 unsigned wordSz, bool useLambdas = false);
+
+  ~WideMemManager() = default;
+
+  Expr bytesToSlotExpr(unsigned bytes) {
+    return m_ctx.alu().si(8 * bytes, g_slotBitWidth);
+  }
+
+  PtrSortTy ptrSort() const {
+    return PtrSortTy(OpSemMemManager::PtrSortTy(),
+                     m_ctx.alu().intTy(g_slotBitWidth));
+  }
+
+  PtrTy salloc(unsigned int bytes, uint32_t align) {
+    assert(isa<AllocaInst>(m_ctx.getCurrentInst()));
+    align = std::max(align, m_alignment);
+    auto region = m_allocator->salloc(bytes, align);
+    assert(region.first - region.second > 0);
+    return PtrTy(mkStackPtr(region.second).getRaw(),
+                 bytesToSlotExpr(region.first - region.second));
```

Comment:
You can add a constructor to `PtrTy` that takes an `Expr` and an `unsigned` and converts the size to `Expr` in the constructor.

Reply:
Let me know if there is another way. m_ctx is not accessible from PtrTy. I will need to pass a reference of WideMemMgr to PtrTy which makes it complicated compared to current `bytesToSlotExpr` impl

Reply:
I thought that `region.first - region.second` is a number and you convert it here to an expr.
Maybe I missed something.

Reply:
Currently I have

```
Expr bytesToSlotExpr(unsigned bytes) {
    return m_ctx.alu().si(bytes, g_slotBitWidth);
  }
```

Reply:
I see. Too much context is then needed to move this into the constructor -- which is what you said. Maybe inlining the function is more readable. I think I got confused by the name.